src/System.Collections.Generic.MultiValueDictionary/src/System/Collections/Generic/MultiValueDictionary.cs

```diff
@@ -0,0 +1,1102 @@
+                            throw new InvalidOperationException((SR.InvalidOperation_EnumEnded));
+                        default:
+                            return current;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Advances the enumerator to the next element of the collection.
+            /// </summary>
+            /// <returns>
+            /// true if the enumerator was successfully advanced to the next element; false if the enumerator has passed the end of the collection.
+            /// </returns>
+            /// <exception cref="T:System.InvalidOperationException">The collection was modified after the enumerator was created. </exception>
+            public bool MoveNext()
+            {
+                if (version != multiValueDictionary.version)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
+                }
+                else if (enumerator.MoveNext())
+                {
+                    current = new KeyValuePair<TKey, IReadOnlyCollection<TValue>>(enumerator.Current.Key, (IReadOnlyCollection<TValue>)enumerator.Current.Value);
+                    state = EnumerationState.During;
+                    return true;
+                }
+                else
+                {
+                    current = default(KeyValuePair<TKey, IReadOnlyCollection<TValue>>);
+                    state = EnumerationState.AfterLast;
+                    return false;
+                }
+            }
+
+            /// <summary>
+            /// Sets the enumerator to its initial position, which is before the first element in the collection.
+            /// </summary>
+            /// <exception cref="T:System.InvalidOperationException">The collection was modified after the enumerator was created. </exception>
+            public void Reset()
+            {
+                if (version != multiValueDictionary.version)
+                    throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
+                enumerator.Dispose();
+                enumerator = multiValueDictionary.dictionary.GetEnumerator();
+                current = default(KeyValuePair<TKey, IReadOnlyCollection<TValue>>);
+                state = EnumerationState.BeforeFirst;
+            }
+
+            /// <summary>
+            /// Frees resources associated with this Enumerator
+            /// </summary>
+            public void Dispose()
+            {
+                enumerator.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// An inner class that functions as a view of an ICollection within a MultiValueDictionary
+        /// </summary>
+        private class InnerCollectionView :
+            ICollection<TValue>,
+            IReadOnlyCollection<TValue>
+        {
+            private TKey key;
+            private ICollection<TValue> collection;
```

Comment:
Why not use the concrete type here? Right now you doing two interface dispatch calls per access to the IROCollection